sgx/epid/src/lib.rs

```diff
@@ -0,0 +1,188 @@
+///
+/// This is used to contact IAS and retrieve the [`SigRL`] data structure.
+impl EpidQuotingEnclave for EpidGroupId {
+    fn for_epid_qe() -> SgxResult<EpidGroupId> {
+        let mut target_info = TargetInfo::default();
+        let mut epid_gid = EpidGroupId::default();
+
+        unsafe { sgx_init_quote(target_info.as_mut(), epid_gid.as_mut()) }.into_result(epid_gid)
+    }
+}
+
+/// Create a new quote from an existing enclave's report.
+pub trait EpidQuoteReport {
+    /// Contact the Quoting Enclave to create a new quote from the given report.
+    ///
+    /// Given a report created inside an enclave using the Quoting Enclave's target info, this
+    /// method will contact the Quoting Enclave, and return a tuple containing the quote and the
+    /// Quoting Enclave's own report, targeting the original enclave.
+    ///
+    /// Therefore, the stages are:
+    ///
+    ///  1. Create a new report inside Enclave A targeting the Quoting Enclave.
+    ///  1. Enclave A exports the report to the untrusted code.
+    ///  1. Untrusted code calls this method to create a new quote, using the given nonce.
+    ///  1. Untrusted code provides the resulting quote and QE enclave's report to enclave A.
+    ///  1. Enclave A verifies the QE enclave's report targets Enclave A.
+    ///  1. Enclave A verifies the QE enclave's report data contains the hash of the nonce and the
+    ///     quote.
+    ///  1. Enclave A verifies the contents of the quote.
+    fn quote(
+        sigrl: &SignatureRevocationList,
+        report: &Report,
+        quote_type: QuoteSign,
+        provider_id: &ProviderId,
+        nonce: &QuoteNonce,
+    ) -> SgxResult<(Quote, Report)>;
+}
+
+impl EpidQuoteReport for Quote {
+    fn quote(
+        sigrl: &SignatureRevocationList,
+        report: &Report,
+        quote_type: QuoteSign,
+        provider_id: &ProviderId,
+        nonce: &QuoteNonce,
+    ) -> SgxResult<(Quote, Report)> {
+        let sigrl_ref = sigrl.as_ref();
+        let mut quote_size = 0u32;
+        unsafe { sgx_calc_quote_size(sigrl_ref.as_ptr(), sigrl_ref.len() as u32, &mut quote_size) }
+            .into_result(())?;
+
+        let mut quote = Quote::with_capacity(quote_size as usize)
+            .expect("SGX SDK requested a quote larger than we are allowed to use");
+
+        let mut qe_report = Report::default();
+
+        unsafe {
+            sgx_get_quote(
+                report.as_ref(),
+                quote_type.into(),
+                provider_id.as_ref(),
+                nonce.as_ref(),
+                sigrl_ref.as_ptr(),
+                sigrl_ref.len() as u32,
+                qe_report.as_mut(),
+                quote.as_mut(),
+                quote_size,
+            )
+        }
+        .into_result((quote, qe_report))
+    }
+}
+
+bitflags! {
+    /// A set of bitflags describing what kind of reprovisioning to perform, if any.
+    pub struct EpidUpdateConfig: u32 {
```

Comment:
I notice that EpidUpdateStatus starts with 1, while this starts with a shift (1 << 1). Just checking if that's intentional.

Reply:
Yeah, bit zero is reserved in the config u32 in the Intel API.